[JSC] PutByOffset and PutClosureVar should interact with DoubleRep directly

Constellation opened the pull request:
#### c29ba63981cc26c709ef1fa8b65770f5d6a0fd0b
<pre>
[JSC] PutByOffset and PutClosureVar should interact with DoubleRep directly
<a href="https://bugs.webkit.org/show_bug.cgi?id=279809">https://bugs.webkit.org/show_bug.cgi?id=279809</a>
<a href="https://rdar.apple.com/136126295">rdar://136126295</a>

Reviewed by Yijia Huang.

When storing DoubleRep to property fields, we need to box them and store them.
On ARM64, we first convert it from FPRReg to GPRReg and do box operation on
the top of that. However this fmov is very costly operation.

Fortunately, on ARM64, we can add / sub integer values on FPRReg (this
is separate from addDouble) so we can box DoubleRep into JSValue format
in FPRReg. And if we can use storeDouble, then we do not need to use
fmov.

This patch extends DFG ValueRepReductionPhase on ARM64, and figuring out
this possibility. And when we found PutByOffset(..., @0) pattern with
appropriate value flows, we convert it to PutByOffset(..., DoubleRep:@0)
and do the above optimization.

We do these optimization as a part of ValueRepReductionPhase so that we
do not need to consider about Object Allocation Sinking&apos;s Double
handling (since Object Allocation Sinking is already done).

* Source/JavaScriptCore/assembler/ARM64Assembler.h:
(JSC::ARM64Assembler::add):
(JSC::ARM64Assembler::adr): Deleted.
(JSC::ARM64Assembler::adrp): Deleted.
(JSC::ARM64Assembler::and_): Deleted.
(JSC::ARM64Assembler::asr): Deleted.
(JSC::ARM64Assembler::asrv): Deleted.
(JSC::ARM64Assembler::b): Deleted.
(JSC::ARM64Assembler::b_cond): Deleted.
(JSC::ARM64Assembler::bfc): Deleted.
(JSC::ARM64Assembler::bfi): Deleted.
(JSC::ARM64Assembler::bfm): Deleted.
(JSC::ARM64Assembler::bfxil): Deleted.
(JSC::ARM64Assembler::bic): Deleted.
(JSC::ARM64Assembler::bl): Deleted.
(JSC::ARM64Assembler::blr): Deleted.
(JSC::ARM64Assembler::br): Deleted.
(JSC::ARM64Assembler::brk): Deleted.
(JSC::ARM64Assembler::isBrk): Deleted.
(JSC::ARM64Assembler::cbnz): Deleted.
(JSC::ARM64Assembler::cbz): Deleted.
(JSC::ARM64Assembler::ccmn): Deleted.
(JSC::ARM64Assembler::ccmp): Deleted.
(JSC::ARM64Assembler::cinc): Deleted.
(JSC::ARM64Assembler::cinv): Deleted.
(JSC::ARM64Assembler::cls): Deleted.
(JSC::ARM64Assembler::clz): Deleted.
(JSC::ARM64Assembler::cmn): Deleted.
(JSC::ARM64Assembler::cmp): Deleted.
(JSC::ARM64Assembler::cneg): Deleted.
(JSC::ARM64Assembler::csel): Deleted.
(JSC::ARM64Assembler::cset): Deleted.
(JSC::ARM64Assembler::csetm): Deleted.
(JSC::ARM64Assembler::csinc): Deleted.
(JSC::ARM64Assembler::csinv): Deleted.
(JSC::ARM64Assembler::csneg): Deleted.
(JSC::ARM64Assembler::eon): Deleted.
(JSC::ARM64Assembler::eor): Deleted.
(JSC::ARM64Assembler::extr): Deleted.
(JSC::ARM64Assembler::hint): Deleted.
(JSC::ARM64Assembler::hlt): Deleted.
(JSC::ARM64Assembler::illegalInstruction): Deleted.
(JSC::ARM64Assembler::isValidLDPImm): Deleted.
(JSC::ARM64Assembler::isValidLDPFPImm): Deleted.
(JSC::ARM64Assembler::ldp): Deleted.
(JSC::ARM64Assembler::ldnp): Deleted.
(JSC::ARM64Assembler::ldr): Deleted.
(JSC::ARM64Assembler::ldr_literal): Deleted.
(JSC::ARM64Assembler::ldrb): Deleted.
(JSC::ARM64Assembler::ldrh): Deleted.
(JSC::ARM64Assembler::ldrsb): Deleted.
(JSC::ARM64Assembler::ldrsh): Deleted.
(JSC::ARM64Assembler::ldrsw): Deleted.
(JSC::ARM64Assembler::ldrsw_literal): Deleted.
(JSC::ARM64Assembler::ldur): Deleted.
(JSC::ARM64Assembler::ldurb): Deleted.
(JSC::ARM64Assembler::ldurh): Deleted.
(JSC::ARM64Assembler::ldursb): Deleted.
(JSC::ARM64Assembler::ldursh): Deleted.
(JSC::ARM64Assembler::ldursw): Deleted.
(JSC::ARM64Assembler::lsl): Deleted.
(JSC::ARM64Assembler::lslv): Deleted.
(JSC::ARM64Assembler::lsr): Deleted.
(JSC::ARM64Assembler::lsrv): Deleted.
(JSC::ARM64Assembler::madd): Deleted.
(JSC::ARM64Assembler::mneg): Deleted.
(JSC::ARM64Assembler::simdQBit): Deleted.
(JSC::ARM64Assembler::encodeLaneAndIndex): Deleted.
(JSC::ARM64Assembler::ins): Deleted.
(JSC::ARM64Assembler::umov): Deleted.
(JSC::ARM64Assembler::smov): Deleted.
(JSC::ARM64Assembler::dupElement): Deleted.
(JSC::ARM64Assembler::dupGeneral): Deleted.
(JSC::ARM64Assembler::fcmeq): Deleted.
(JSC::ARM64Assembler::fcmgt): Deleted.
(JSC::ARM64Assembler::fcmge): Deleted.
(JSC::ARM64Assembler::vectorNot): Deleted.
(JSC::ARM64Assembler::sizeForIntegralSIMDOp): Deleted.
(JSC::ARM64Assembler::sizeForFloatingPointSIMDOp): Deleted.
(JSC::ARM64Assembler::cmeq): Deleted.
(JSC::ARM64Assembler::cmeqz): Deleted.
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::loadVector):
(JSC::MacroAssemblerARM64::add64):
(JSC::MacroAssemblerARM64::sub64):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::add64):
(JSC::MacroAssemblerX86_64::sub64):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:
* Source/JavaScriptCore/dfg/DFGValueRepReductionPhase.cpp:
(JSC::DFG::ValueRepReductionPhase::convertValueRepsToDouble):
* Source/JavaScriptCore/ftl/FTLCommonValues.cpp:
(JSC::FTL::CommonValues::initializeConstants):
* Source/JavaScriptCore/ftl/FTLCommonValues.h:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::boxDoubleAsDouble):
(JSC::FTL::DFG::LowerDFGToB3::unboxDoubleAsDouble):
(JSC::FTL::DFG::LowerDFGToB3::compilePutByOffset):
(JSC::FTL::DFG::LowerDFGToB3::compileMultiPutByOffset):
(JSC::FTL::DFG::LowerDFGToB3::compilePutGlobalVariable):
(JSC::FTL::DFG::LowerDFGToB3::compilePutClosureVar):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/JSCJSValue.h:
* Source/JavaScriptCore/runtime/PureNaN.h:

Canonical link: <a href="https://commits.webkit.org/283786@main">https://commits.webkit.org/283786@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6615f587f769d9eb1c2665f2bf26e21a921ee73b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67401 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/46780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/71449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/18538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/54578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/18329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/71449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/18538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70468 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/54578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/71449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/54578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/16896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/60516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/54578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/73148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/66646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/11360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/18329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/73148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/11395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/73148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/88415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10228 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/88415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/43662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/44848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/43403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/JSCJSValue.h:
```diff
@@ -479,6 +479,8 @@ class JSValue {
     // with a 15-bit pattern within the range 0x0002..0xFFFC.
     static constexpr size_t DoubleEncodeOffsetBit = 49;
     static constexpr int64_t DoubleEncodeOffset = 1ll << DoubleEncodeOffsetBit;
+    static constexpr double EncodedPNaN = std::bit_cast<double>(PNaNAsBits + DoubleEncodeOffset);
```
This is not used, right?

Constellation replied:
Yes, removed.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] PutByOffset and PutClosureVar should interact with DoubleRep directly
https://bugs.webkit.org/show_bug.cgi?id=279809
rdar://136126295

Reviewed by Yijia Huang.

When storing DoubleRep to property fields, we need to box them and store them.
On ARM64, we first convert it from FPRReg to GPRReg and do box operation on
the top of that. However this fmov is very costly operation.

Fortunately, on ARM64, we can add / sub integer values on FPRReg (this
is separate from addDouble) so we can box DoubleRep into JSValue format
in FPRReg. And if we can use storeDouble, then we do not need to use
fmov.

This patch extends DFG ValueRepReductionPhase on ARM64, and figuring out
this possibility. And when we found PutByOffset(..., @0) pattern with
appropriate value flows, we convert it to PutByOffset(..., DoubleRep:@0)
and do the above optimization.

We do these optimization as a part of ValueRepReductionPhase so that we
do not need to consider about Object Allocation Sinking's Double
handling (since Object Allocation Sinking is already done).

* Source/JavaScriptCore/assembler/ARM64Assembler.h:
(JSC::ARM64Assembler::add):
(JSC::ARM64Assembler::adr): Deleted.
(JSC::ARM64Assembler::adrp): Deleted.
(JSC::ARM64Assembler::and_): Deleted.
(JSC::ARM64Assembler::asr): Deleted.
(JSC::ARM64Assembler::asrv): Deleted.
(JSC::ARM64Assembler::b): Deleted.
(JSC::ARM64Assembler::b_cond): Deleted.
(JSC::ARM64Assembler::bfc): Deleted.
(JSC::ARM64Assembler::bfi): Deleted.
(JSC::ARM64Assembler::bfm): Deleted.
(JSC::ARM64Assembler::bfxil): Deleted.
(JSC::ARM64Assembler::bic): Deleted.
(JSC::ARM64Assembler::bl): Deleted.
(JSC::ARM64Assembler::blr): Deleted.
(JSC::ARM64Assembler::br): Deleted.
(JSC::ARM64Assembler::brk): Deleted.
(JSC::ARM64Assembler::isBrk): Deleted.
(JSC::ARM64Assembler::cbnz): Deleted.
(JSC::ARM64Assembler::cbz): Deleted.
(JSC::ARM64Assembler::ccmn): Deleted.
(JSC::ARM64Assembler::ccmp): Deleted.
(JSC::ARM64Assembler::cinc): Deleted.
(JSC::ARM64Assembler::cinv): Deleted.
(JSC::ARM64Assembler::cls): Deleted.
(JSC::ARM64Assembler::clz): Deleted.
(JSC::ARM64Assembler::cmn): Deleted.
(JSC::ARM64Assembler::cmp): Deleted.
(JSC::ARM64Assembler::cneg): Deleted.
(JSC::ARM64Assembler::csel): Deleted.
(JSC::ARM64Assembler::cset): Deleted.
(JSC::ARM64Assembler::csetm): Deleted.
(JSC::ARM64Assembler::csinc): Deleted.
(JSC::ARM64Assembler::csinv): Deleted.
(JSC::ARM64Assembler::csneg): Deleted.
(JSC::ARM64Assembler::eon): Deleted.
(JSC::ARM64Assembler::eor): Deleted.
(JSC::ARM64Assembler::extr): Deleted.
(JSC::ARM64Assembler::hint): Deleted.
(JSC::ARM64Assembler::hlt): Deleted.
(JSC::ARM64Assembler::illegalInstruction): Deleted.
(JSC::ARM64Assembler::isValidLDPImm): Deleted.
(JSC::ARM64Assembler::isValidLDPFPImm): Deleted.
(JSC::ARM64Assembler::ldp): Deleted.
(JSC::ARM64Assembler::ldnp): Deleted.
(JSC::ARM64Assembler::ldr): Deleted.
(JSC::ARM64Assembler::ldr_literal): Deleted.
(JSC::ARM64Assembler::ldrb): Deleted.
(JSC::ARM64Assembler::ldrh): Deleted.
(JSC::ARM64Assembler::ldrsb): Deleted.
(JSC::ARM64Assembler::ldrsh): Deleted.
(JSC::ARM64Assembler::ldrsw): Deleted.
(JSC::ARM64Assembler::ldrsw_literal): Deleted.
(JSC::ARM64Assembler::ldur): Deleted.
(JSC::ARM64Assembler::ldurb): Deleted.
(JSC::ARM64Assembler::ldurh): Deleted.
(JSC::ARM64Assembler::ldursb): Deleted.
(JSC::ARM64Assembler::ldursh): Deleted.
(JSC::ARM64Assembler::ldursw): Deleted.
(JSC::ARM64Assembler::lsl): Deleted.
(JSC::ARM64Assembler::lslv): Deleted.
(JSC::ARM64Assembler::lsr): Deleted.
(JSC::ARM64Assembler::lsrv): Deleted.
(JSC::ARM64Assembler::madd): Deleted.
(JSC::ARM64Assembler::mneg): Deleted.
(JSC::ARM64Assembler::simdQBit): Deleted.
(JSC::ARM64Assembler::encodeLaneAndIndex): Deleted.
(JSC::ARM64Assembler::ins): Deleted.
(JSC::ARM64Assembler::umov): Deleted.
(JSC::ARM64Assembler::smov): Deleted.
(JSC::ARM64Assembler::dupElement): Deleted.
(JSC::ARM64Assembler::dupGeneral): Deleted.
(JSC::ARM64Assembler::fcmeq): Deleted.
(JSC::ARM64Assembler::fcmgt): Deleted.
(JSC::ARM64Assembler::fcmge): Deleted.
(JSC::ARM64Assembler::vectorNot): Deleted.
(JSC::ARM64Assembler::sizeForIntegralSIMDOp): Deleted.
(JSC::ARM64Assembler::sizeForFloatingPointSIMDOp): Deleted.
(JSC::ARM64Assembler::cmeq): Deleted.
(JSC::ARM64Assembler::cmeqz): Deleted.
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::loadVector):
(JSC::MacroAssemblerARM64::add64):
(JSC::MacroAssemblerARM64::sub64):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::add64):
(JSC::MacroAssemblerX86_64::sub64):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGHeapLocation.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGHeapLocation.h:
* Source/JavaScriptCore/dfg/DFGStoreBarrierInsertionPhase.cpp:
* Source/JavaScriptCore/dfg/DFGValueRepReductionPhase.cpp:
(JSC::DFG::ValueRepReductionPhase::convertValueRepsToDouble):
* Source/JavaScriptCore/ftl/FTLCommonValues.cpp:
(JSC::FTL::CommonValues::initializeConstants):
* Source/JavaScriptCore/ftl/FTLCommonValues.h:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::boxDoubleAsDouble):
(JSC::FTL::DFG::LowerDFGToB3::unboxDoubleAsDouble):
(JSC::FTL::DFG::LowerDFGToB3::compilePutByOffset):
(JSC::FTL::DFG::LowerDFGToB3::compileMultiPutByOffset):
(JSC::FTL::DFG::LowerDFGToB3::compilePutGlobalVariable):
(JSC::FTL::DFG::LowerDFGToB3::compilePutClosureVar):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/JSCJSValue.h:
* Source/JavaScriptCore/runtime/PureNaN.h:

Canonical link: https://commits.webkit.org/283786@main
